taskweaver/ces/environment.py

```diff
@@ -94,92 +98,253 @@ def pre_start_kernel(
         kwargs: Any,
     ) -> tuple[KernelManager, str, str]:
         env: Optional[Dict[str, str]] = kwargs.get("env")
+
         km, kernel_name, kernel_id = super().pre_start_kernel(kernel_name, kwargs)
-        if env is not None and "CONNECTION_FILE" in env:
-            km.connection_file = env["CONNECTION_FILE"]
+        if env is not None:
+            if "CONNECTION_FILE" in env:
+                km.connection_file = env["CONNECTION_FILE"]
+            if "JUPYTER_MANUAL_PORTS" in env:
+                # otherwise, the ports will be assigned automatically
+                km.cache_ports = False
+            if "JUPYTER_SHELL_PORT" in env:
+                km.shell_port = int(env["JUPYTER_SHELL_PORT"])
+            if "JUPYTER_STDIN_PORT" in env:
+                km.stdin_port = int(env["JUPYTER_STDIN_PORT"])
+            if "JUPYTER_CONTROL_PORT" in env:
+                km.control_port = int(env["JUPYTER_CONTROL_PORT"])
+            if "JUPYTER_HB_PORT" in env:
+                km.hb_port = int(env["JUPYTER_HB_PORT"])
+            if "JUPYTER_IOPUB_PORT" in env:
+                km.iopub_port = int(env["JUPYTER_IOPUB_PORT"])
+            if "JUPYTER_KERNEL_IP" in env:
+                km.ip = env["JUPYTER_KERNEL_IP"]
         return km, kernel_name, kernel_id
 
 
+class EnvMode(enum.Enum):
+    SubProcess = "subprocess"
+    InsideContainer = "inside_container"
+    OutsideContainer = "outside_container"
+
+
 class Environment:
     def __init__(
         self,
         env_id: Optional[str] = None,
         env_dir: Optional[str] = None,
+        env_mode: Optional[EnvMode] = EnvMode.SubProcess,
+        port_start_inside_container: Optional[int] = 12345,
     ) -> None:
         self.session_dict: Dict[str, EnvSession] = {}
         self.id = get_id(prefix="env") if env_id is None else env_id
         self.env_dir = env_dir if env_dir is not None else os.getcwd()
+        self.mode = env_mode
+        if self.mode == EnvMode.SubProcess or self.mode == EnvMode.InsideContainer:
+            self.multi_kernel_manager = TaskWeaverMultiKernelManager(
+                default_kernel_name="taskweaver",
+                kernel_spec_manager=KernelSpecProvider(),
+            )
+            if self.mode == EnvMode.InsideContainer:
+                file_handler = logging.FileHandler("env.log")
+                file_handler.setLevel(logging.DEBUG)
+                file_handler.setFormatter(formatter)
+                logger.addHandler(file_handler)
+
+        elif self.mode == EnvMode.OutsideContainer:
+            try:
+                import docker
+                import docker.errors
+            except ImportError:
+                raise ImportError("docker package is required for container-based kernel.")
```

Comment:
Do we need to have docker as dependency in requirements? Maybe we can remove it from there if we have import check here?

Reply:
The problem is that we may need to set `container` mode as default. Let's discuss it.

Reply:
not as default